Docker Build Error

The docker build method described in the repositories throws the following error.
Is there any workaround for this?

I am testing on Ubuntu 22.04. Poetry and all components are installed and working fine.

```
~/iso15118 (master) » sudo make build                                                                                                                                                
cd iso15118/shared/pki; ./create_certs.sh -v iso-2
Password used is: '12345'
read EC key
writing EC key
Signature ok
subject=CN = V2GRootCA, O = Switch, C = UK, DC = V2G
Getting Private key
read EC key
writing EC key
Signature ok
subject=CN = CPOSubCA1, O = Switch, C = UK, DC = V2G
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = CPOSubCA2, O = Switch, C = UK, DC = V2G
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = SECCCert, O = Switch, C = UK, DC = CPO
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = OEMRootCA, O = Switch, C = UK, DC = OEM
Getting Private key
read EC key
writing EC key
Signature ok
subject=CN = OEMSubCA1, O = Switch, C = UK, DC = OEM
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = OEMSubCA2, O = Switch, C = UK, DC = OEM
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = OEMProvCert, O = Switch, C = UK, DC = OEM
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = MORootCA, O = Switch, C = UK, DC = MO
Getting Private key
read EC key
writing EC key
Signature ok
subject=CN = PKI-Ext_CRT_MO_SUB1_VALID, O = Switch, C = UK, DC = MO
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = PKI-Ext_CRT_MO_SUB2_VALID, O = Switch, C = UK, DC = MO
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = UKSWI123456789A, O = Switch, C = UK, DC = MO
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = ProvSubCA1, O = Switch, C = UK, DC = CPS
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = ProvSubCA2, O = Switch, C = UK, DC = CPS
Getting CA Private Key
read EC key
writing EC key
Signature ok
subject=CN = CPS Leaf, O = Switch, C = UK, DC = CPS
Getting CA Private Key
'template.Dockerfile' -> 'iso15118/evcc/Dockerfile'
'template.Dockerfile' -> 'iso15118/secc/Dockerfile'
docker-compose build
redis uses an image, skipping
Building secc
Sending build context to Docker daemon  30.24MB
Step 1/18 : FROM python:3.10.0-buster as build
 ---> b47b4cc944d6
Step 2/18 : WORKDIR /usr/src/app
 ---> Using cache
 ---> c7093a479e83
Step 3/18 : ENV PYTHONFAULTHANDLER=1   PYTHONUNBUFFERED=1   PYTHONHASHSEED=random   PIP_NO_CACHE_DIR=1   PIP_DISABLE_PIP_VERSION_CHECK=1   PIP_DEFAULT_TIMEOUT=100   POETRY_VERSION=1.1.11   VIRTUALENV_PIP=21.2.1   MYPY_VERSION=0.930
 ---> Using cache
 ---> 3052b10cd20c
Step 4/18 : RUN pip install "poetry==$POETRY_VERSION" "mypy==$MYPY_VERSION"
 ---> Using cache
 ---> 2e72d4652c00
Step 5/18 : COPY .coveragerc mypy.ini .flake8 poetry.lock pyproject.toml ./
 ---> Using cache
 ---> 208ea18bbae2
Step 6/18 : RUN sed -i 's/secc/secc/g' pyproject.toml
 ---> Using cache
 ---> f2aa97bba272
Step 7/18 : RUN poetry install --no-interaction --no-ansi
 ---> Running in e002b0470aa8

  RuntimeError

  The Poetry configuration is invalid:
    - Additional properties are not allowed ('group' was unexpected)


  at /usr/local/lib/python3.10/site-packages/poetry/core/factory.py:43 in create_poetry
       39│             message = ""
       40│             for error in check_result["errors"]:
       41│                 message += "  - {}\n".format(error)
       42│
    →  43│             raise RuntimeError("The Poetry configuration is invalid:\n" + message)
       44│
       45│         # Load package
       46│         name = local_config["name"]
       47│         version = local_config["version"]
The command '/bin/sh -c poetry install --no-interaction --no-ansi' returned a non-zero code: 1
ERROR: Service 'secc' failed to build : Build failed
make: *** [Makefile:59: build] Error 1
```

--- a/Makefile
+++ b/Makefile
@@ -56,6 +56,8 @@ build: .generate_v2_certs
 	@# This conversion is required, otherwise we wouldn't be able to spawn the evcc start script.
 	@# @ is used as a separator and allows us to escape '/', so we can substitute the '/' itself
 	@sed -i.bkp 's@/secc/g@/evcc/g@g' iso15118/evcc/Dockerfile
+	@# Add a delay on EVCC to give SECC time to start up 
+	@sed -i 's/CMD\ \/venv\/bin\/iso15118/CMD\ echo\ "Waiting for 5 seconds to start EVCC"\ \&\&\ sleep\ 5\ \&\&\ \/venv\/bin\/iso15118/g' iso15118/evcc/Dockerfile	
 	docker-compose build
 
 # Run using dev env vars

--- a/Makefile
+++ b/Makefile
@@ -57,7 +57,7 @@ build: .generate_v2_certs
 	@# @ is used as a separator and allows us to escape '/', so we can substitute the '/' itself
 	@sed -i.bkp 's@/secc/g@/evcc/g@g' iso15118/evcc/Dockerfile
 	@# Add a delay on EVCC to give SECC time to start up 
-	@sed -i 's/CMD\ \/venv\/bin\/iso15118/CMD\ echo\ "Waiting for 5 seconds to start EVCC"\ \&\&\ sleep\ 5\ \&\&\ \/venv\/bin\/iso15118/g' iso15118/evcc/Dockerfile	
+	@sed -i'.bkp' -e 's@CMD /venv/bin/iso15118@CMD echo "Waiting for 5 seconds to start EVCC" \&\& sleep 5 \&\& /venv/bin/iso15118@g' iso15118/evcc/Dockerfile
 	docker-compose build
 
 # Run using dev env vars

--- a/README.md
+++ b/README.md
@@ -10,7 +10,8 @@ The primary dependencies to install the project are the following:
 > - Linux
 >
 >   - MacOS is not fully supported -- see "IPv6 Warning" below
->   - Other non-Linux operating systems are not supported*
+>   - Windows is supported using [WSL](https://learn.microsoft.com/en-us/windows/wsl/install).
+>   - Other non-Linux operating systems are not supported
 >
 > - Poetry [^3]
 > - Python >= 3.9

--- a/iso15118/secc/transport/tcp_server.py
+++ b/iso15118/secc/transport/tcp_server.py
@@ -81,10 +81,14 @@ class TCPServer(asyncio.Protocol):
             port = self.port_tls
             ssl_context = get_ssl_context(True)
             server_type = "TLS"
-        # Initialise socket for IPv6 TCP packets
-        # Address family (determines network layer protocol, here IPv6)
-        # Socket type (stream, determines transport layer protocol TCP)
-        while True:
+
+        MAX_RETRIES: int = 3
+        BACK_OFF_SECONDS: float = 0.5
+
+        for i in range(MAX_RETRIES):
+            # Initialise socket for IPv6 TCP packets
+            # Address family (determines network layer protocol, here IPv6)
+            # Socket type (stream, determines transport layer protocol TCP)
             sock = socket.socket(family=socket.AF_INET6, type=socket.SOCK_STREAM)
 
             # Allows address to be reused
@@ -97,11 +101,16 @@ class TCPServer(asyncio.Protocol):
             # TCP packets
             try:
                 sock.bind(self.full_ipv6_address)
+                break
             except OSError as e:
-                logger.info(f"{e} on {server_type} server. Retrying...")
-                await asyncio.sleep(0.5)
-                continue
-            break
+                # Once the max amount of retries has been reached, reraise the exception
+                if i == MAX_RETRIES - 1:
+                    raise e
+                else:
+                    logger.info(f"{e} on {server_type} server. Retrying...")
+                    await asyncio.sleep(BACK_OFF_SECONDS)
+                    continue
+
         server = await asyncio.start_server(
             # The client_connected_cb callback, which is the __call__ method of
             # this class) is called whenever a new client connection is

--- a/iso15118/secc/transport/tcp_server.py
+++ b/iso15118/secc/transport/tcp_server.py
@@ -84,7 +84,10 @@ class TCPServer(asyncio.Protocol):
 
         MAX_RETRIES: int = 3
         BACK_OFF_SECONDS: float = 0.5
-
+        # Note: When the socket is being created inside a container,
+        # sometimes the network interface is not ready yet and the binding
+        # process fails the first time.
+        # Therefore, a wait-and-retry block has been added.
         for i in range(MAX_RETRIES):
             # Initialise socket for IPv6 TCP packets
             # Address family (determines network layer protocol, here IPv6)

--- a/iso15118/secc/transport/tcp_server.py
+++ b/iso15118/secc/transport/tcp_server.py
@@ -84,18 +84,24 @@ class TCPServer(asyncio.Protocol):
         # Initialise socket for IPv6 TCP packets
         # Address family (determines network layer protocol, here IPv6)
         # Socket type (stream, determines transport layer protocol TCP)
-        sock = socket.socket(family=socket.AF_INET6, type=socket.SOCK_STREAM)
+        while True:
+            sock = socket.socket(family=socket.AF_INET6, type=socket.SOCK_STREAM)
 
-        # Allows address to be reused
-        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            # Allows address to be reused
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
 
-        self.full_ipv6_address = await get_link_local_full_addr(port, self.iface)
-        self.ipv6_address_host = self.full_ipv6_address[0]
+            self.full_ipv6_address = await get_link_local_full_addr(port, self.iface)
+            self.ipv6_address_host = self.full_ipv6_address[0]
 
-        # Bind the socket to the IP address and port for receiving
-        # TCP packets
-        sock.bind(self.full_ipv6_address)
-
+            # Bind the socket to the IP address and port for receiving
+            # TCP packets
+            try:
+                sock.bind(self.full_ipv6_address)
+            except OSError as e:
+                logger.info(f"{e} on {server_type} server. Retrying...")
+                await asyncio.sleep(0.5)
+                continue
+            break
         server = await asyncio.start_server(
             # The client_connected_cb callback, which is the __call__ method of
             # this class) is called whenever a new client connection is

--- a/template.Dockerfile
+++ b/template.Dockerfile
@@ -9,7 +9,7 @@ ENV PYTHONFAULTHANDLER=1 \
   PIP_NO_CACHE_DIR=1 \
   PIP_DISABLE_PIP_VERSION_CHECK=1 \
   PIP_DEFAULT_TIMEOUT=100 \
-  POETRY_VERSION=1.1.11 \
+  POETRY_VERSION=1.3.2 \
   VIRTUALENV_PIP=21.2.1 \
   MYPY_VERSION=0.930
 
@@ -46,6 +46,7 @@ RUN /venv/bin/pip install dist/*.whl
 
 # Replace with in-container cert generation DevOps#2664
 COPY --from=build /usr/src/app/iso15118/shared/pki/ /usr/src/app/iso15118/shared/pki/
+COPY --from=build /usr/src/app/iso15118/shared/examples/evcc/iso15118_2/ /usr/src/app/iso15118/shared/examples/evcc/iso15118_2/
 
 RUN /venv/bin/pip install aiofile
 # This will run the entrypoint script defined in the pyproject.toml

